src/main/org/pantsbuild/jarjar/misplaced/MisplacedClassProcessor.java

```diff
@@ -0,0 +1,73 @@
+package org.pantsbuild.jarjar.misplaced;
+
+import java.io.IOException;
+import org.objectweb.asm.ClassReader;
+import org.pantsbuild.jarjar.util.ContextualJarProcessor;
+import org.pantsbuild.jarjar.util.EntryStruct;
+import org.pantsbuild.jarjar.util.ProcessingContext;
+
+/**
+ * Processor that handles classes whose fully-qualified names do not match their jar entry location.
+ */
+public abstract class MisplacedClassProcessor extends ContextualJarProcessor {
+
+  public MisplacedClassProcessor(ProcessingContext context) {
+    super(context);
+  }
+
+  /**
+   * Handles a class that was found at a location that does not match its fully qualified classname.
+   * @param classStruct The jar EntryStruct for the class.
+   * @param className The actual fully-qualified classname read from the binary data.
+   */
+  public abstract void handleMisplacedClass(EntryStruct classStruct, String className);
+
+  /**
+   * Whether any JarTransformers transform and relocate the struct to its "proper" location based on
+   * its package name, or just skip over it.
+   * @return True if the struct should be transformed, false otherwise.
+   */
+  public abstract boolean shouldTransform();
+
+  /**
+   * Returns whether the entry should be included in the output.
+   * @return true if it should be kept, false if it should be omitted.
+   */
+  public abstract boolean shouldKeep();
+
+  @Override public boolean process(EntryStruct struct)
+      throws IOException {
+    if (!struct.name.endsWith(".class")) return true;
+
+    String originalClassName = null;
+    ClassReader reader;
+    try {
+      reader = new ClassReader(struct.data);
+      originalClassName = reader.getClassName() + ".class";
+    } catch (Exception e) {
+      System.err.println("Unable to read classname from bytecode in " + struct.name);
+      System.err.println("Shading is therefore impossible, so this entry will be skipped.");
+      System.err.println(e.getClass().getName() + ": " + e.getMessage());
+      getContext().excludeFromTranforming(struct);
+      return true;
+    }
+    if (!originalClassName.equals(struct.name)) {
+      if (getContext().isVerbose()) {
+        System.err.println(formatMessage(struct, originalClassName));
+      }
+      handleMisplacedClass(struct, originalClassName);
+      if (!shouldTransform()) {
+        getContext().excludeFromTranforming(struct);
+      }
+      return shouldKeep();
+    }
+    return true;
+  }
+
+  protected String formatMessage(EntryStruct classStruct, String className) {
```

Comment:
Perhaps `formatMisplcaedMessage` or otherwise javadoc for subclasses to know what sort of message this generates.  I'm happy enough though with a more specific name.